src/MiddlewareListener.php

```diff
@@ -117,51 +116,85 @@ public function onDispatch(MvcEvent $event)
     /**
      * Create a middleware pipe from the array spec given.
      *
-     * @param mixed[] $middlewaresToBePiped
-     *      List of string ids for middlewares in container, instances of MiddlewareInterface
-     *      or callable middlewares
-     * @return MiddlewarePipe
+     * @param string|MiddlewareInterface|RequestHandlerInterface|PipeSpec $middlewarePipeSpec
      * @throws InvalidMiddlewareException
      */
     private function createPipeFromSpec(
-        ContainerInterface $serviceLocator,
-        ResponseInterface $responsePrototype,
-        array $middlewaresToBePiped
-    ) {
+        ContainerInterface $container,
+        $middlewarePipeSpec
+    ): RequestHandlerInterface {
+        if (is_string($middlewarePipeSpec)) {
+            $middlewarePipeSpec = $this->middlewareFromContainer($container, $middlewarePipeSpec);
+        }
+        if ($middlewarePipeSpec instanceof RequestHandlerInterface) {
+            return $middlewarePipeSpec;
+        }
+
         $pipe = new MiddlewarePipe();
-        $pipe->setResponsePrototype($responsePrototype);
-        foreach ($middlewaresToBePiped as $middlewareToBePiped) {
+        if ($middlewarePipeSpec instanceof MiddlewareInterface) {
+            $pipe->pipe($middlewarePipeSpec);
+            return $pipe;
+        }
```

Comment:
Wouldn't it be cleaner if a new `PipeSpec` instance was created which will be handled further down anyways?

Reply:
I thought of that, yes